_posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd

```diff
@@ -0,0 +1,492 @@
+
+## Purpose and scope of this post
+
+Here, I introduce [`escnn`](https://github.com/QUVA-Lab/escnn), a PyTorch extension that implements forms of group equivariance for CNNs operating on the plane or in (3d) space. The library is used in various, amply illustrated research papers; it is appropriately documented; and it comes with introductory notebooks both relating the math and exercising the code. Why, then, not just refer to the [first notebook](https://github.com/QUVA-Lab/escnn/blob/master/examples/introduction.ipynb), and immediately start using it for some experiment?
+
+In fact, this post should -- as quite a few texts I've written -- be regarded as an introduction to an introduction. To me, this topic seems anything but easy, for various reasons. Of course, there's the math. But as so often in machine learning, you don't need to go to great depths to be able to apply an algorithm correctly. So if not the math itself, what generates the difficulty? For me, it's two things.
+
+First, to map my understanding of the mathematical concepts to the terminology used in the library, and from there, to correct use and application. Expressed schematically: We have a concept A, which figures (among other concepts) in technical term (or object class) B. What does my understanding of A tell me about how object class B is to be used correctly? More importantly: How do I use it to best attain my goal C? -- This first difficulty I'll address in a very pragmatic way. I'll neither dwell on mathematical details, nor try to establish the links between A, B, and C in detail. Instead, I'll present the characters[^2] in this story by asking what they're good for.
+
+[^2]: If you have some background on representations: No, not *those* characters ...
+
+Second -- and this will be of relevance to just a subset of readers -- the topic of group equivariance, particularly as applied to image processing, is one where visualizations can be of tremendous help. The quaternity[^3] of conceptual explanation, math, code, and visualization can, together, produce an understanding of emergent-seeming quality... if, and only if, all of these explanation modes "work" for you. (Or if, in an area, a mode that does not wouldn't contribute that much anyway.) Here, it so happens that from what I saw, several papers have excellent visualizations[^4], and the same holds for some lecture slides and accompanying notebooks[^5]. But for those among us with limited spatial-imagination capabilities -- e.g., people with [Aphantasia](https://en.wikipedia.org/wiki/Aphantasia) -- these illustrations, intended to help, can be very hard to make sense of themselves. If you're not one of these, I totally recommend checking out the resources linked in the above footnotes. This text, though, will try to make the best possible use of verbal explanation to introduce the concepts involved, the library, and how to use it.
```

Comment:
> This text, though, will try to make the best possible use of verbal explanation to introduce the concepts involved, the library, and how to use it.

OK, good. Now I know what you're about to be telling me! There are a lot of caveats and asides in these first few paragraphs, so it's _very_ helpful to end with a clear declarative statement at the end.

Reply:
Yeah, you're right :-) I know, there's many caveats & all ... I also know I usually don't go into details about what makes things hard, but in this post, I really wanted to, subjective and personal though it may be...